includes/classes/SupportMonitor/DBQueryMonitor.php

```diff
@@ -0,0 +1,173 @@
+<?php
+/**
+ * Database Queries Monitor. A submodule of Support Monitor to report heavy SQL queries executed on staging.
+ *
+ * @since  x.x
+ * @package 10up-experience
+ */
+
+namespace TenUpExperience\SupportMonitor;
+
+/**
+ * DBQueryMonitor class
+ */
+class DBQueryMonitor {
+
+	/**
+	 * The transient name
+	 *
+	 * This transient stores all the queries logged.
+	 */
+	const TRANSIENT_NAME = 'tenup_experience_db_queries';
+
+	/**
+	 * Setup module
+	 */
+	public function setup() {
+		$production_environment = Monitor::instance()->get_setting( 'production_environment' );
+		if ( 'no' === $production_environment || apply_filters( 'tenup_experience_log_heavy_queries', false ) ) {
+			add_filter( 'query', [ $this, 'maybe_log_query' ] );
+		}
+	}
+
+	/**
+	 * Conditionally log a query
+	 *
+	 * Get all potential heavy queries (CREATE, ALTER, etc.) and store it,
+	 * ignoring transients and options by default.
+	 *
+	 * @param string $query The SQL query.
+	 * @return string
+	 */
+	public function maybe_log_query( $query ) {
+		global $wpdb;
+
+		if ( ! preg_match( '/^\s*(create|alter|truncate|drop|insert|delete|update|replace)\s/i', $query ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, 'transient_' ) ) {
+			return $query;
+		}
+
+		if ( false !== strpos( $query, $wpdb->options ) ) {
+			return $query;
+		}
+
+		if ( apply_filters( 'tenup_experience_log_query', true, $query ) ) {
+			$this->log_query( $query );
+		}
+
+		return $query;
+	}
+
+	/**
+	 * Get the logged queries. Also removes from the transient all queries logged for more than 7 days.
+	 *
+	 * @return array
+	 */
+	public function get_report() {
+		$date_limit = strtotime( '7 days ago' );
+
+		$stored_queries = array_filter(
+			(array) get_transient( self::TRANSIENT_NAME ),
+			function ( $query_date ) use ( $date_limit ) {
+				return strtotime( $query_date ) > $date_limit;
+			},
+			ARRAY_FILTER_USE_KEY
+		);
+
+		set_transient( self::TRANSIENT_NAME, $stored_queries );
+
+		return $stored_queries;
+	}
+
+	/**
+	 * Log/store a SQL query
+	 *
+	 * Queries are stored like:
+	 *
+	 *    'YYYY-MM-DD' => [
+	 *        'a5be998feee8968155052c4d332a7223' => [ // md5 of file:line:query
+	 *            'query' => 'ALTER TABLE wp_my_db_heavy_plugin CHANGE COLUMN `id` id BIGINT(20) UNSIGNED NOT NULL AUTO_INCREMENT',
+	 *            'file' => '/var/www/html/wp-content/plugins/my-db-heavy-plugin/my-db-heavy-plugin.php',
+	 *            'line' => 53,
+	 *            'count' => 1,
```

Comment:
@felipeelia Curious what the 'count' attribute will be doing. It seems to me we'd be able to get this via the order the data is stored. Not a big deal, but I wanted to point it out.

Reply:
@bengreeley my intent with it was just to avoid duplicate entries on the same day. So, for example, we had a plugin that tried to alter the same table several times. Instead of storing the same queries, the code would be just increasing that count. Makes sense?

Reply:
Yep makes sense to me